[Enhancement] Custom symbols for start/end edge segments instead of squares, custom whitespace control

n1kk opened the pull request:
Hi, this may look like a big change but it's really two files, the rest rest are minor adjustments to the test cases.

I wanted to achieve this look, so I added ability to specify symbols for left and right edges of prompts instead of default whitespace with square bg. I could've added it to first line prefix but then it wouldn't be colored.

![screen shot 2018-11-17 at 7 50 59 pm](https://user-images.githubusercontent.com/5442968/48664129-d4c3bf80-eaa2-11e8-9391-fc913ba7fcb9.png)

While playing with style symbols I realized that I needed to control right and left whitespace of segments separately, so I added ws control for all cases, left/right prompt, left/right side of default or custom segment, and a middle one between the icon and the content.
Using some symbols adds too much space, but removing ws altogether makes it to cramped or overlapped. 

![screen shot 2018-11-17 at 8 15 41 pm](https://user-images.githubusercontent.com/5442968/48664363-38032100-eaa6-11e8-8186-edab384ac636.png)
![screen shot 2018-11-17 at 8 15 53 pm](https://user-images.githubusercontent.com/5442968/48664369-3df90200-eaa6-11e8-913c-6e1ae6732145.png)
![screen shot 2018-11-17 at 8 16 26 pm](https://user-images.githubusercontent.com/5442968/48664371-40f3f280-eaa6-11e8-9493-5a8a041d0fbe.png)
![screen shot 2018-11-17 at 8 16 35 pm](https://user-images.githubusercontent.com/5442968/48664372-43564c80-eaa6-11e8-9515-e8998e95d5cc.png)

Here is an example of the left prompt where ws present only on the left (except first seg) and in the middle.

![screen shot 2018-11-17 at 7 52 09 pm](https://user-images.githubusercontent.com/5442968/48664227-333d6d80-eaa4-11e8-9370-3bd285dba1d2.png)
![screen shot 2018-11-17 at 8 09 31 pm](https://user-images.githubusercontent.com/5442968/48664271-b8288700-eaa4-11e8-96a2-9004c2da3244.png)

Looks more compact and cleaner. No extra whitespace and no icon overlapping. And now it is 
 possible to remove whitespace between the icon and the content. Works in single and multi line mode, and there's more ways to mix and match styles now. You can make right prompt more compact to be more out of the way.

![screen shot 2018-11-17 at 9 08 25 pm](https://user-images.githubusercontent.com/5442968/48664809-2a9d6500-eaad-11e8-975d-7a6e3c6b9ef9.png)

In total I've added three new methods to utilities.zsh and modified/refactored three methods in default.p9k. The rest of changes files are new specs and fixes to existing specs where color codes and whitespace just moved around a bit, no structural changes. I covered all new utilities and functionality with test cases.

Short summary:
- refactored variables a bit to make sure that variables that do not have visible content stay empty (do not include escape codes)
- added utility for finding defined value to make it possible to easily specify variable override order
- added custom left symbol for first segment of PROMPT (with custom ws) and same for last segment of RPROMPT
- wrapped parts in isolated vars, allowed me to remove some extra if cases, make code more compact
- added whitespace controll for each segment for left, right and middle

Everything is optional and doesn't change anything if vars are not defined.

Examples of the new config variables can be seen in `test/core/prompt.spec`.

All tests are passing locally. Docker didn't work for me, after starting it I try to enter into `p9k` dir and it just hogs cpu and hangs. TravisCI passed all test except for zsh v5.2 and lower there was one error in git test of extra whitespace before the dir path. [travis report](https://travis-ci.org/n1kk/powerlevel9k/builds/456391244)
```
testGitDirClobber
ASSERT:expected:<%K{001} %F{000}✘ /tmp/powerlevel9k-test/test-dotfiles  master ✚ ? %k%F{001}%f > but was:<%K{001} %F{000}✘  /tmp/powerlevel9k-test/test-dotfiles  master ✚ ? %k%F{001}%f >
```

Hope this is a nice addition to customization control and stylizing.

I can write some docs for this, I just need to know where exactly to do it.

dritter commented:
Hi @n1kk !
This looks awesome! I'll need some time to review this.

FYI I fixed the tests, but Travis seems to confuse your `next` branch with ours.. Strange. Anyway, the tests from [your build](https://travis-ci.org/n1kk/powerlevel9k/builds/456734071?utm_source=github_status&utm_medium=notification) are green.

// requesting review from @onaforeignshore

n1kk commented:
Hey @dritter, thanks for the fix, I've found and fixed one my mistake also. But then check failed because of new vagrant test cases, so I fetched upstream and fixed those too. Hope now all check will pass.

Will be waiting for a review from @onaforeignshore.

bhilburn commented:
Very cool work! Thanks for leading the review & merge effort, @dritter!

@n1kk - Excited to see this upstreamed! `=)`

n1kk commented:
Hey @dritter, thanks for the fixes, I've merged your changes into this PR and pulled upstream to sync potential test conflicts, there were a couple in battery spec. All are resolved and test are passing.

dritter commented:
Awesome. Thanks.

While debugging #1051 I stumbled upon a regression (I was still on your branch): I you use the newline segment and configured away the Icon for it, we end up in `__p9k_left_prompt_segment` in the line that joins the whitespaces `echo -n "$(p9k::join_by "${middle_ws}" "${visual_identifier}" "${content}")${right_ws}"`. The trick was even if `echo -n` was called, it respected `$'\n'` and printed a newline, but with a function that joins together the parameters, newlines seem to get stripped:
![bildschirmfoto 2018-12-12 um 22 33 30](https://user-images.githubusercontent.com/1544760/49900194-2d963600-fe5e-11e8-83b6-14c55c6e9360.png)

Trying to fix this in the next days..

dritter commented:
Thx @n1kk
